wicket-core/src/main/java/org/apache/wicket/pageStore/DiskPageStore.java

```diff
@@ -0,0 +1,717 @@
+
+	private static final String INDEX_FILE_NAME = "DiskPageStoreIndex";
+
+	/**
+	 * A cache that holds all page stores.
+	 */
+	private static final ConcurrentMap<String, DiskPageStore> DISK_STORES = new ConcurrentHashMap<>();
+
+	private final String applicationName;
+
+	private final ISerializer serializer;
+
+	private final Bytes maxSizePerSession;
+
+	private final NestedFolders folders;
+
+	private final ConcurrentMap<String, DiskData> diskDatas;
+
+	/**
+	 * Create a store that supports {@link SerializedPage}s only.
+	 * 
+	 * @param applicationName
+	 *            name of application
+	 * @param fileStoreFolder
+	 *            folder to store to
+	 * @param maxSizePerSession
+	 *            maximum size per session
+	 * 
+	 * @see SerializingPageStore
+	 */
+	public DiskPageStore(String applicationName, File fileStoreFolder, Bytes maxSizePerSession)
+	{
+		this(applicationName, fileStoreFolder, maxSizePerSession, null);
+	}
+
+	/**
+	 * Create a store to disk.
+	 * 
+	 * @param applicationName
+	 *            name of application
+	 * @param fileStoreFolder
+	 *            folder to store to
+	 * @param maxSizePerSession
+	 *            maximum size per session
+	 * @param serializer
+	 *            for serialization of pages
+	 */
+	public DiskPageStore(String applicationName, File fileStoreFolder, Bytes maxSizePerSession,
+		ISerializer serializer)
+	{
+		this.applicationName = Args.notNull(applicationName, "applicationName");
+		this.folders = new NestedFolders(new File(fileStoreFolder, applicationName + "-filestore"));
+		this.maxSizePerSession = Args.notNull(maxSizePerSession, "maxSizePerSession");
+		this.serializer = serializer; // optional
+
+		this.diskDatas = new ConcurrentHashMap<>();
+
+		try
+		{
+			if (folders.getBase().exists() || folders.getBase().mkdirs())
+			{
+				loadIndex();
+			}
+			else
+			{
+				log.warn("Cannot create file store folder for some reason.");
+			}
+		}
+		catch (SecurityException e)
+		{
+			throw new WicketRuntimeException(
+				"SecurityException occurred while creating DiskPageStore. Consider using a non-disk based IPageStore implementation. "
+					+ "See org.apache.wicket.Application.setPageManagerProvider(IPageManagerProvider)",
+				e);
+		}
+
+		if (DISK_STORES.containsKey(applicationName))
+		{
+			throw new IllegalStateException(
+				"Store for application with key '" + applicationName + "' already exists.");
+		}
+		DISK_STORES.put(applicationName, this);
+	}
+
+	/**
+	 * Versioning is supported if a serializer was provided to the constructor.
+	 */
+	@Override
+	public boolean supportsVersioning()
+	{
+		return serializer != null;
+	}
+
+	@Override
+	public void destroy()
+	{
+		log.debug("Destroying...");
+		saveIndex();
+
+		DISK_STORES.remove(applicationName);
+
+		log.debug("Destroyed.");
+	}
+
+	@Override
+	public boolean canBeAsynchronous(IPageContext context)
+	{
+		// session attribute must be added here *before* any asynchronous calls
+		// when session is no longer available
```

Comment:
AsynchronousPageStore uses a task, it's named PendingAdd now:
It does some magic so that the code being executed on another thread can still access the IPageContext, e.g. the stored session attributes *although* the session is actually no longer accessible from there.